Proposals/NNNN-calendar-recurrence-rule.md

```diff
@@ -0,0 +1,466 @@
+        public var hours: [Int]?
+        /// On which days of the week the event should occur
+        public var weekdays: [Weekday]?
+        /// On which days in the month the event should occur
+        /// - 1 signifies the first day of the month.
+        /// - Negative values point to a day counted backwards from the last day
+        ///   of the month
+        /// This field is unused when `frequency` is `.weekly`.
+        public var daysOfTheMonth: [Int]?
+        /// On which days of the year the event may occur.
+        /// - 1 signifies the first day of the year.
+        /// - Negative values point to a day counted backwards from the last day
+        ///   of the year
+        /// This field is unused when `frequency` is any of `.daily`, `.weekly`,
+        /// or `.monthly`.
+        public var daysOfTheYear: [Int]?
+        /// On which months the event should occur.
+        /// - 1 is the first month of the year (January in Gregorian calendars)
+        public var months: [Month]?
+        /// On which weeks of the year the event should occur.
+        /// - 1 is the first week of the year. `calendar.minimumDaysInFirstWeek`
+        ///   defines which week is considered first.
+        /// - Negative values refer to weeks if counting backwards from the last
+        ///   week of the year. -1 is the last week of the year.
+        /// This field is unused when `frequency` is other than `.yearly`.
+        public var weeks: [Int]?
+        /// Which occurrences within every interval should be returned
+        public var setPositions: [Int]?
+
+        public init(calendar: Calendar,
+                    frequency: Frequency,
+                    interval: Int = 1,
+                    end: End = .never,
+                    matchingPolicy: Calendar.MatchingPolicy = .nextTimePreservingSmallerComponents,
+                    repeatedTimePolicy: RepeatedTimePolicy = .onlyFirst,
+                    months: [Month]? = nil,
+                    daysOfTheYear: [Int]? = nil,
+                    daysOfTheMonth: [Int]? = nil,
+                    weeks: [Int]? = nil,
+                    weekdays: [Weekday]? = nil,
+                    hours: [Int]? = nil,
+                    minutes: [Int]? = nil,
+                    seconds: [Int]? = nil,
+                    setPositions: [Int]? = nil) -> Self
+
+        /// Find recurrences of the given date
+        ///
+        /// The calculations are implemented according to RFC-5545 and RFC-7529.
+        ///
+        /// - Parameter start: the date at which to start search for recurrences
+        /// - Parameter range: a range of dates which to search for recurrences.
+        ///   If `nil`, return all recurrences of the event.
+        /// - Returns: a sequence of dates conforming to the recurrence rule, in
+        ///   the given `range`. An empty sequence if the rule doesn't match any
+        ///   dates.
+        public func recurrences(of start: Date,
```

Comment:
I can see use cases for other expand methods:

* Expand N instances (from start) - i.e. use a count instead of a range
* Expand N instances (after instance D) - e.g., give me the next instance (N=1) instance after D

Note sure if these are things you want to include.

Reply:
I think both can be achieved using the existing API without a performance cost:

```swift
var count = 0
for date in recurrenceRule.recurrences(of: .now) while count < 10 {
    count += 1
    ...
}
```
The range argument is more necessary because it allows to skip calculating recurrences until the start of the range.